src/renderer/mesh_loader.zig

```diff
@@ -1,18 +1,192 @@
+    const num_indices: u32 = @as(u32, @intCast(primitive.indices.?.count));
+
+    var indices = std.ArrayList(rt.IndexType).init(arena);
+    var positions = std.ArrayList([3]f32).init(arena);
+    var normals = std.ArrayList([3]f32).init(arena);
+    var tangents = std.ArrayList([4]f32).init(arena);
+    var uvs = std.ArrayList([2]f32).init(arena);
+    defer indices.deinit();
+    defer positions.deinit();
+    defer normals.deinit();
+    defer tangents.deinit();
+    defer uvs.deinit();
+
+    // Indices.
+    {
+        try indices.ensureTotalCapacity(indices.items.len + num_indices);
+
+        const accessor = primitive.indices.?;
+        const buffer_view = accessor.buffer_view.?;
+
+        assert(accessor.stride == buffer_view.stride or buffer_view.stride == 0);
+        assert(accessor.stride * accessor.count == buffer_view.size);
+        assert(buffer_view.buffer.data != null);
+
+        const data_addr = @as([*]const u8, @ptrCast(buffer_view.buffer.data)) +
+            accessor.offset + buffer_view.offset;
+
+        if (accessor.stride == 1) {
+            assert(accessor.component_type == .r_8u);
+            const src = @as([*]const u8, @ptrCast(data_addr));
+            var i: u32 = 0;
+            while (i < num_indices) : (i += 1) {
+                indices.appendAssumeCapacity(src[i]);
+            }
+        } else if (accessor.stride == 2) {
+            assert(accessor.component_type == .r_16u);
+            const src = @as([*]const u16, @ptrCast(@alignCast(data_addr)));
+            var i: u32 = 0;
+            while (i < num_indices) : (i += 1) {
+                indices.appendAssumeCapacity(src[i]);
+            }
+        } else if (accessor.stride == 4) {
+            assert(accessor.component_type == .r_32u);
+            const src = @as([*]const u32, @ptrCast(@alignCast(data_addr)));
+            var i: u32 = 0;
+            while (i < num_indices) : (i += 1) {
+                indices.appendAssumeCapacity(src[i]);
+            }
+        } else {
+            unreachable;
+        }
+    }
+
+    // Attributes.
+    {
+        const attributes = primitive.attributes[0..primitive.attributes_count];
+        for (attributes) |attrib| {
+            const accessor = attrib.data;
+            assert(accessor.component_type == .r_32f);
+
+            const buffer_view = accessor.buffer_view.?;
+            assert(buffer_view.buffer.data != null);
+
+            assert(accessor.stride == buffer_view.stride or buffer_view.stride == 0);
+            assert(accessor.stride * accessor.count == buffer_view.size);
+
+            const data_addr = @as([*]const u8, @ptrCast(buffer_view.buffer.data)) +
+                accessor.offset + buffer_view.offset;
+
+            if (attrib.type == .position) {
+                assert(accessor.type == .vec3);
+                const slice = @as([*]const [3]f32, @ptrCast(@alignCast(data_addr)))[0..num_vertices];
```

Comment:
With complex/long casts like this my general preference is to split it into two lines, but this is fine.